# Model Evaluation Exercises

In [5]:
import pandas as pd
import numpy as np
from env import get_connection
import seaborn as sns
import matplotlib.pyplot as plt
from pydataset import data
import os
from scipy import stats

#turn off the pink warning boxes
import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn import metrics

import acquire
import prepare

from env import get_connection
pd.options.display.max_columns = None

%matplotlib inline

1

Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- In the context of this problem, what is a false positive?

- In the context of this problem, what is a false negative? 

- How would you describe this model?

I would determine here that:
- Actual dog and pred dog is True Negative
- Actual cat pred dog is the False Negative
- Actual cat pred cat is True Negative
- Actual dog pred cat is False Positive



2 

You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

- Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

- Use the predictions dataset and pandas to help answer the following questions:

In [ ]:
codey = pd.read_csv('codey.csv')

In [4]:
codey.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


    -- In this case we would be looking for Precision
    -- I would use the first model for this since it has the highest precision when it comes to finding ducks with defects

In [7]:
print(metrics.classification_report(codey['actual'], codey['model1']))

              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200



In [9]:
print(metrics.classification_report(codey['actual'], codey['model2']))

              precision    recall  f1-score   support

      Defect       0.10      0.56      0.17        16
   No Defect       0.94      0.56      0.70       184

    accuracy                           0.56       200
   macro avg       0.52      0.56      0.44       200
weighted avg       0.87      0.56      0.66       200



In [8]:
print(metrics.classification_report(codey['actual'], codey['model3']))

              precision    recall  f1-score   support

      Defect       0.13      0.81      0.23        16
   No Defect       0.97      0.53      0.69       184

    accuracy                           0.56       200
   macro avg       0.55      0.67      0.46       200
weighted avg       0.90      0.56      0.65       200



- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


     -- I would evaluate this using the Recall rate, since we are trying to determine the accuracy of our capturing of positives, and avoiding false positives
     -- I would use model 3 for this
     
     

3

You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [11]:
paws = pd.read_csv('gives_you_paws.csv')

paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

- In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?


In [25]:
paws['actual'].value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [26]:
3254/5000

0.6508

Baseline accuracy is 65.08%

In [27]:
paws['model1'].value_counts(), 2937/5000

(dog    2937
 cat    2063
 Name: model1, dtype: int64,
 0.5874)

Model 1 accuracy is 58.74%

In [28]:
paws['model2'].value_counts(), 3212/5000

(cat    3212
 dog    1788
 Name: model2, dtype: int64,
 0.6424)

Model 2 accuracy is 64.24%

In [29]:
paws['model3'].value_counts(), 2508/5000

(dog    2508
 cat    2492
 Name: model3, dtype: int64,
 0.5016)

Model 3 accuracy is 50.16%

In [30]:
paws['model4'].value_counts(), 4253/5000

(dog    4253
 cat     747
 Name: model4, dtype: int64,
 0.8506)

Model 4 accuracy is 85.06%

- Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?


    -- I would use model 4 for the team that deals with dog pictures, as that model more accurately determines when a picture is a dog.

- Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?


    -- I would use model 2 for the team that uses cat picutres, as that model more accurately guesses when a picture is a cat.

Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- sklearn.metrics.accuracy_score
- sklearn.metrics.precision_score
- sklearn.metrics.recall_score
- sklearn.metrics.classification_report

- Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend? -- With sklearn methods


    -- I would use the 1st model for this, since this model is tied with model 2 for precision but the recall has a higher probability

In [31]:
print(metrics.classification_report(paws['actual'], paws['model1']))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [32]:
print(metrics.classification_report(paws['actual'], paws['model2']))

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [33]:
print(metrics.classification_report(paws['actual'], paws['model3']))

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [34]:
print(metrics.classification_report(paws['actual'], paws['model4']))

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000



- Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend? -- With sklearn methods


    -- I would use model 4 for the cat team, since it has the highest precision at guessing which picture is a cat